### PR TITLE
[Feature] Support batch invariance on ROCm

### PR DESCRIPTION
**Pull request by mawong-amd**
## Purpose
This PR enables broad-based batch invariant inference on ROCm, keyed by `VLLM_BATCH_INVARIANT=1`.

The benefits of this PR are however not limited to ROCm.
1. All-reduce and reduce-scatter are supported via `CustomAllReduce` and `CustomReduceScatter`. The 1stage algorithm is already batch-invariant, and the 2stage algorithm can be made batch-invariant by a one-line change. **This makes the entirety of custom all-reduce/reduce-scatter batch-invariant, including on CUDA**, which likely can lead to performance improvements there. However, this code path is disabled on CUDA pending tuning in a follow-up PR.
    - A corresponding fallback via all-gather and fixed-order sum is added that supports all platforms.
    - On ROCm gfx950, custom all-reduce / reduce-scatter have been measured to be on par with if not faster than the RCCL implementations.

2. Triton kernels have been added for batch-invariant softmax / scaled FP8 GEMM.
3. Fixes a longstanding bug where `torch.mm`/`torch.bmm`/`torch.addmm` were not batch-invariant under Inductor (because Inductor lowers to the `.out` variants) has been fixed.

## Testing
Numerous E2E runs were carried out on a variety of models (DeepSeek R1, Llama-4-Maverick, GPT-OSS-120B, including quantized variants) across a variety of parallelism strategies (TP, PP, DP, EP, DCP) and logprobs examined to confirm batch invariance. The E2E runs also confirm that gsm8k accuracy is not degraded by batch invariance.

The batch-invariance test groups on AMD CI (both gfx942 and gfx950) pass.

New tests have been added to the aforementioned test groups:

| module | GPUs | covers |
|---|---|---|
| `test_mx_linear_batch_invariant.py` | 1 (gfx950) | the native MX linear kernels, including the four AITER MXFP4 entry points |
| `test_moe_row_permutation_batch_invariant.py` | 1 | the fused-MoE expert GEMM against row order within an expert |
| `test_tp_all_reduce_batch_invariant.py` | 4 | all-reduce across token counts spanning the protocol thresholds |
| `test_tp_reduce_scatter_batch_invariant.py` | 4 | the same for reduce-scatter |

The four upstream modules the first commit un-skips —
`test_batch_invariance`, `test_rms_norm_batch_invariant`,
`test_matmul_batch_invariant`, `test_online_batch_invariance` — cover the rest of
what this touches and now actually run on ROCm.

Run on 8x gfx950 (MI355), ROCm 7.2, `VLLM_WORKER_MULTIPROC_METHOD=spawn`, from
`tests/`:

| command | result |
|---|---|
| `pytest v1/determinism/test_matmul_batch_invariant.py` | 26 passed, 17s |
| `pytest v1/determinism/test_rms_norm_batch_invariant.py` | 475 passed, 2m35s |
| `pytest v1/determinism/test_mx_linear_batch_invariant.py` | 21 passed, 17s |
| `pytest v1/determinism/test_moe_row_permutation_batch_invariant.py` | 23 passed, 13s |
| `pytest v1/determinism/test_tp_all_reduce_batch_invariant.py -m 'distributed(num_gpus=4)'` | 6 passed, 4m00s |
| `pytest v1/determinism/test_tp_reduce_scatter_batch_invariant.py -m 'distributed(num_gpus=4)'` | 6 passed, 3m59s |
| `pytest v1/determinism/test_batch_invariance.py` | 6 passed, 1 skipped, 5m03s |

## AI assistance
Written with AI assistance (Claude). Every line has been reviewed by the submitter, who can defend the change end to end.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by mawong-amd** [comment]
cc @yewentao256 as the code owner for batch invariance.

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work! Could we firstly shrink the diff? Ideally < 1k LOC

**Comment by mawong-amd** [comment]
> Thanks for the work! Could we firstly shrink the diff? Ideally < 1k LOC

I'll move the all-reduce / reduce-scatter tests to a follow-up PR. I've already withheld a bunch of tests hardening batch invariance verification, so they'll fit in there, and this PR can be scoped to not adding any testing capability beyond matching what's currently in upstream.

I'll keep the MX linear tests in this PR though: they verify that MXFP4/MXFP8 matmul through AITER are currently batch invariant (with some config tweaks) but that could always move out from under us.

Shrinking beyond this will start cutting into features (batch-invariant custom all-reduce/reduce-scatter, FP8, MXFP4, MXFP8), so I'd like to avoid that if possible. In terms of LoC, non-test LoC is just under 900 and the bulk of that is the batch-invariant scaled FP8 GEMM and softmax Triton kernels, followed by the custom all-reduce buffer size changes under BI.
In principle those could help all platforms, but I haven't wired them in this PR because I haven't tested performance there vs NCCL with the pinned tree all-reduce algorithms, although I imagine custom all-reduce will be faster. If I'm not mistaken, batch-invariant reduce-scatter doesn't yet exist so this is a pure feature gain.

**Comment by mawong-amd** [comment]
All CI tests green currently, including the 3 Batch Invariance test mirrors on AMD CI ([A100-MI250](https://buildkite.com/vllm/amd-ci/builds/12069/list?sid=01a00193-a6ed-4a00-bda0-d40474656415&tab=output), [H100-MI300](https://buildkite.com/vllm/amd-ci/builds/12069/list?sid=01a00193-a75d-4084-bb8d-73128e36350f&tab=output), and [B200-MI355](https://buildkite.com/vllm/amd-ci/builds/12069/list?sid=01a00193-a793-4d3a-bf48-beaececef0e1&tab=output)). The full test suite caught an existing issue where `linear_batch_invariant` added `bias` to the matmul in the native dtype rather than in the FP32 accumulator, which is less accurate than `torch.nn.functional.linear`; this has been fixed.

**Review by AndreasKaratzas — DISMISSED** [review]
LGTM
@yewentao256 PTAL

File: vllm/model_executor/kernels/linear/__init__.py [hunk 1]
```diff
@@ -680,6 +681,11 @@ def init_fp8_linear_kernel(
     )
 
     if activation_quant_key.scale.group_shape.is_per_group():
+        if envs.VLLM_BATCH_INVARIANT and current_platform.is_rocm():
+            # AiterFp8BlockScaledMMKernel is first in the ROCm priority list and
+            # is not batch invariant; the Triton block kernel is.
+            force_kernel = TritonFp8BlockScaledMMKernel  # type: ignore[assignment]
```
Comment (yewentao256) [inline vllm/model_executor/kernels/linear/__init__.py:687]: Are you sure blcok fp8 here is batch invariant?
https://github.com/vllm-project/vllm/blob/32d925a152446fdecc828c1d8c6373beb8f36b34/vllm/model_executor/layers/quantization/utils/fp8_utils.py#L938-L949

File: vllm/model_executor/layers/fused_moe/experts/ocp_mx_emulation_moe.py [hunk 2]
```diff
@@ -196,7 +196,10 @@ def apply(
             global_num_experts=global_num_experts,
             expert_map=expert_map,
             a1q_scale=None,
-            a2_scale=None,
+            # The second GEMM quantizes from this argument, not from the quant
+            # config: dropping it takes a statically-calibrated scheme down
+            # scaled_fp8_quant's dynamic path.
+            a2_scale=a2_scale,
```
Comment (yewentao256) [inline vllm/model_executor/layers/fused_moe/experts/ocp_mx_emulation_moe.py]:
```suggestion
            a2_scale=a2_scale,
```

File: vllm/model_executor/layers/batch_invariant.py [hunk 3]
```diff
@@ -53,6 +64,7 @@ def matmul_kernel_persistent(
     stride_bn,
     stride_cm,
     stride_cn,
+    stride_bias,
```
Comment (yewentao256) [inline vllm/model_executor/layers/batch_invariant.py:68]: Why we need this stride_bias?

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work! I think there are some updateds needed before merge

Please also take a look at these AI generated comments

```bash
The new BatchInvariantFP8ScaledMMLinearKernel is not consistently wired. Fp8LinearMethod.apply() only allows CUTLASS to use the direct FP8 path; otherwise it still dequantizes to BF16. Some ModelOpt/Quark paths may call the new kernel directly, but the common FP8/online paths do not. Please either wire it consistently and add direct coverage, or avoid forcing/maintaining an implementation those paths bypass.

scaled_mm_kernel_persistent substantially duplicates the existing portable triton_scaled_mm. Could we fix its use_heuristic=False path, use a fixed tile configuration under BI, and move bias addition before the output cast? Unless the persistent scheduler has a demonstrated performance requirement, reusing the existing kernel would remove a significant amount of code.

The collective changes should either restore their tests or move to a follow-up PR. The final diff no longer contains the all-reduce/reduce-scatter tests referenced by the implementation, and the current AMD CI runs with TP=1, so these paths are not directly exercised. In particular, DCP/custom-AR buffer sizing is a performance optimization rather than a correctness requirement and could be split out.

Some smaller cleanup opportunities:
_softmax_kernel mostly duplicates _log_softmax_kernel; a compile-time output-mode flag could share the reduction logic.
tests/v1/determinism/utils.py::bits() is unused.
test_scale_spread_exposes_reordering tests the sensitivity of randomized test data rather than product behavior; a deterministic adversarial input would be simpler and more robust.
The MXFP4 correctness test should reuse the existing reference helpers and likely live in the kernel correctness suite.

I would keep the core MXFP4/MXFP8 batch-invariance sweeps because they exercise ROCm-specific kernels with no direct CUDA equivalent. However, their shared comparison logic and some cases could be consolidated.
```

File: vllm/model_executor/kernels/linear/__init__.py (hunk 1 above)
Comment (mawong-amd) [reply vllm/model_executor/kernels/linear/__init__.py:687]: Good catch! Didn't detect it because testing hardware had no tuned config/only swept over configs with `BLOCK_SIZE_K` that did not affect the accumulation order.
Fixed by forcing the default config under `VLLM_BATCH_INVARIANT=1`.

File: vllm/model_executor/layers/batch_invariant.py (hunk 3 above)
Comment (mawong-amd) [reply vllm/model_executor/layers/batch_invariant.py]: One real-world input that would cause issues:
```
T = torch.randn(8, 4)
B1 = T[:, 0:1].t()   # shape (1,8), stride (1, 4)
B2 = B1.squeeze(0)   # shape (8,), stride (4,)
B3 = B1.reshape(-1)   # shape (8,), stride (4,)
```
`B1` / `B2` / `B3` can be passed as a valid bias to `torch.addmm` and to our `addmm` overrides. Without `stride_bias` we would read a wrong value for every entry other than the first. And there's no easy way to get rid of it via lightweight operations like reshaping/squeezing: we'd have to use `.contiguous()`.

It's worth noting that adding `stride_bias` doesn't hurt in the usual case (where it's = 1): Triton specializes some integer arguments (including 0 and 1) into `tl.constexpr`, so we don't actually do the multiply in VALU because it's optimized away before codegen.

Comment (mawong-amd) [reply vllm/model_executor/layers/batch_invariant.py:68]: One real-world input that would cause issues:
```
T = torch.randn(8, 4)
B1 = T[:, 0:1].t()   # shape (1,8), stride (1, 4)
B2 = B1.squeeze(0)   # shape (8,), stride (4,)
B3 = B1.reshape(-1)   # shape (8,), stride (4,)
```
`B1` / `B2` / `B3` can be passed as a valid bias to `torch.addmm` and to our `addmm` overrides, which eventually land here. Without `stride_bias` we would read a wrong value for every entry other than the first. And there's no easy way to get rid of it via lightweight operations like reshaping/squeezing: we'd have to use `.contiguous()`.

It's worth noting that adding `stride_bias` doesn't change what's executed on GPU in the usual case (where it's = 1): Triton specializes some integer arguments (including 0 and 1) into `tl.constexpr`, so we don't actually do the multiply in VALU because it's optimized away before codegen.

**Comment by mawong-amd** [comment]
> ```The new BatchInvariantFP8ScaledMMLinearKernel is not consistently wired. Fp8LinearMethod.apply() only allows CUTLASS to use the direct FP8 path; otherwise it still dequantizes to BF16. Some ModelOpt/Quark paths may call the new kernel directly, but the common FP8/online paths do not. Please either wire it consistently and add direct coverage, or avoid forcing/maintaining an implementation those paths bypass.```
>
> ```scaled_mm_kernel_persistent substantially duplicates the existing portable triton_scaled_mm. Could we fix its use_heuristic=False path, use a fixed tile configuration under BI, and move bias addition before the output cast? Unless the persistent scheduler has a demonstrated performance requirement, reusing the existing kernel would remove a significant amount of code.```

Addressed together by deleting `scaled_mm_kernel_persistent` and adding a lightweight LinearKernel class `TritonFP8ScaledMMLinearKernel` that calls into `triton_scaled_mm` and is wired into `Fp8LinearMethod.apply()`. It turns out `triton_scaled_mm` is already batch invariant by construction: an existing test was lightly repurposed (`test_fp8_scaled_mm_batch_invariant_fixed_config`) to check this. `triton_scaled_mm` was also lightly modified to do the bias addition before output cast to reduce rounding error (also matching CUTLASS here) and while there a one-line change fixed the `use_heuristic=False` path.


> ```The collective changes should either restore their tests or move to a follow-up PR. The final diff no longer contains the all-reduce/reduce-scatter tests referenced by the implementation, and the current AMD CI runs with TP=1, so these paths are not directly exercised. In particular, DCP/custom-AR buffer sizing is a performance optimization rather than a correctness requirement and could be split out.```

Agreed and done.


> ```_softmax_kernel mostly duplicates _log_softmax_kernel; a compile-time output-mode flag could share the reduction logic.```

Both `_softmax` and `_log_softmax` have been rewritten to use the same basic Triton kernel, which returns one or the other keyed on a `tl.constexpr` flag.


> ```tests/v1/determinism/utils.py::bits() is unused.```
> ```The MXFP4 correctness test should reuse the existing reference helpers and likely live in the kernel correctness suite.```

Both have been removed. The MXFP4 correctness tests are already achieved by `test_aiter_fp4_gemm_skinny_shapes` in                                            `tests/kernels/quantization/test_rocm_mxfp4.py`.


> ```test_scale_spread_exposes_reordering tests the sensitivity of randomized test data rather than product behavior; a deterministic adversarial input would be simpler and more robust.```

`set_random_seed` before `_mxfp4_weights` or `_mxfp8_weights` already means the input should be deterministic: what this test does is ensure that the input is indeed adversarial. It is a positive control for all the sweeps after it.
Put another way, if we got rid of the test without changing anything else, the test suite would have exactly the same testing power as now. What this test guards against is someone adjusting `SCALE_SPREAD` in the future so that tests pass even while not testing anything (because they structurally can't fail). The presence of it at least serves as a signal that in some regimes, particularly batch invariance, simply testing with e.g. default `torch.randn` initialization is not sufficient. The alternative would be a few lines of comments explaining why the weight initializations in the module are done the way they are, but that wouldn't provide information as to how to test whether the initializations are vacuous.



With the latest changes, the total non-test LoC diff has been cut to +514/-61; test and CI LoC diff are increased to +1050/-45.

File: vllm/v1/attention/backends/mla/prefill/flash_attn.py [hunk 4]
```diff
@@ -414,13 +414,13 @@ def _flash_attn_varlen_diff_headdims(
...
             kwargs["output_scale"] = output_scale
+            if envs.VLLM_BATCH_INVARIANT:
+                kwargs["num_splits"] = 1
         else:
             # ROCm leverages the upstream flash_attn, which takes a parameter
             # called "return_attn_probs" instead of return_softmax_lse
             kwargs["return_attn_probs"] = return_softmax_lse
             assert out is None and output_scale is None
-        if envs.VLLM_BATCH_INVARIANT:
-            kwargs["num_splits"] = 1
```
Comment (yewentao256) [inline vllm/v1/attention/backends/mla/prefill/flash_attn.py:423]: We don't pass num_splits=1 in Rocm? Please check here

File: tests/v1/determinism/test_mx_linear_batch_invariant.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    a, b = _reference_operands(fmt, n, k)
+
+    forward = torch.einsum("mk,nk->mn", a, b).to(torch.bfloat16)
+    reverse = torch.einsum("mk,nk->mn", a.flip(-1), b.flip(-1)).to(torch.bfloat16)
+    assert not torch.equal(forward, reverse), (
+        f"reversing the K order leaves every {fmt} output bit unchanged at "
+        f"SCALE_SPREAD={SCALE_SPREAD} (N={n}, K={k}): the fp32 accumulation is "
+        f"exact for these operands, so the batch-invariance sweep over this "
+        f"shape cannot fail"
+    )
```
Comment (yewentao256) [inline tests/v1/determinism/test_mx_linear_batch_invariant.py:198]:
```suggestion
```

This test seems not that meaningful

**Review by yewentao256 — COMMENTED** [review]
Thanks for the update

Please take a look if there are some tests not that meaningful, ideally all diff <1k LOC

Please take a look at these AI generated comments

```bash
1. Online FP8 wiring:
TritonFP8ScaledMMLinearKernel is now used by the regular FP8 path, but the online per-tensor and PTPC paths still only allow CUTLASS to execute FP8 directly under batch-invariant mode. On ROCm, init_fp8_linear_kernel() selects the Triton kernel, but these two apply() methods immediately bypass it and dequantize to BF16. Could we include the Triton kernel in these direct paths, ideally through a shared capability check, and add online-path coverage?

2. Inductor .out regression test:
The .out registrations fix the Inductor issue described in the PR, but the corresponding regression tests were removed. The remaining matmul tests call the helpers directly and do not exercise torch.compile or the .out dispatch. Could we retain one minimal compiled test covering mm/addmm/bmm across batch sizes?

3. Block-FP8 coverage:
Taking the default block-FP8 config under batch-invariant mode looks correct, but the Qwen E2E test does not cross an (N, K) configuration where BLOCK_SIZE_K changes with M. Could we add a direct regression test using one of those tuned shapes, embedding the same row in launches on opposite sides of the BK=64/128 transition and checking bitwise equality?

4. Bias-rounding coverage:
Moving bias addition before the output cast changes triton_scaled_mm semantics for all callers, but the existing test reference still casts the matmul result before adding bias and uses a loose tolerance that masks this difference. Please update the reference to add bias in FP32 before the final cast and add a targeted rounding-sensitive case.
```
